### PR TITLE
API/ENH: relabel method

### LINKED ISSUE #14829: API: allow DataFrame.rename to take a list-like of colums
This is closely related to #12392, but I think separate issue.  Proposal would be to be able to pass a list-like to `DataFrame.rename` to make method-chaining easier.  I think it would also be consistent with #11980 (Series rename)

```python
df = pd.DataFrame({'a': [1,2], 'b': [3, 4]})

# make this
df = df.rename(columns=['j', 'k'])

# equivalent to 
df.columns = ['j', 'k']
```

### PR DESCRIPTION
**Pull request by chris-b1**
Alt to #15029

 - [x] closes #14829
 - [ ] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

small example
```python
In [7]: df = pd.DataFrame({'a': [1,2], 'b': [3, 4]})

In [8]: df.relabel(columns=['J', 'K'])
Out[8]: 
   J  K
0  1  3
1  2  4
```

### CODE REVIEW

**Comment by jreback** [comment]
how would this work with groupby? (just thinking out loud)

**Comment by chris-b1** [comment]
Setup:
```python
df = pd.DataFrame({'key': [1, 1, 2], 'value': [1, 2, 3]})
gb = df.groupby('key')
```

I can't think of a meaningful answer for what `gb.relabel` would do?  You would of course be able to do:
```python
In [17]: gb.sum()
Out[17]: 
     value
key       
1        3
2        3

In [18]: gb.sum().relabel(columns=['newvalue'])
Out[18]: 
     newvalue
key          
1           3
2           3
```

File: pandas/core/generic.py [hunk 1]
```diff
@@ -688,6 +688,92 @@ def f(x):
...
+
+        .. versionadded:: 0.20.0
+
+        Parameters
+        ----------
+        %(axes)s : list-like, optional
+            Labels to construct new axis from - number of labels
+            must match the length of the existing axis
+        copy : boolean, default True
+            Also copy underlying data
```
Comment (jreback) [inline pandas/core/generic.py]: I know we have this flag (``copy``) on ``.rename`` as well. I think these are confusing, though not sure what a better API is for these. (and ``copy=True, inplace=True`` is pretty much meaningless.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -688,6 +688,92 @@ def f(x):
...
+            Whether to return a new %(klass)s. If True then value of copy is
+            ignored.
+
+        Returns
+        -------
+        relabeled : %(klass)s (new object)
+
+        See Also
+        --------
+        pandas.NDFrame.rename
```
Comment (jreback) [inline pandas/core/generic.py]: prob best to add ``pandas.Series.rename, pandas.DataFrame.rename, pandas.Panel.rename`` here instead

File: pandas/core/generic.py (hunk 1 above)
Comment (chris-b1) [inline pandas/core/generic.py]: Yeah, not sure - I agree it's confusing.  I'd be tempted to deprecate `copy` altogether, though if you know what you're doing I suppose it's useful.

File: doc/source/indexing.rst [hunk 3]
```diff
@@ -1586,6 +1586,23 @@ If you create an index yourself, you can just assign it to the ``index`` field:
 
    data.index = index
 
+.. versionadded:: 0.20.0
+
+Alternatively, the :meth:`~DataFrame.relabel` can be used to assign new
```
Comment (jreback) [inline doc/source/indexing.rst]: are these underlined? or is that the diff?

Comment (chris-b1) [inline doc/source/indexing.rst]: It's the diff - no clue why it looks like that though

**Comment by jreback** [comment]
what does SQL, Spark, R call this rename/relabel API?

not that we need to follow, but consistency is not a bad thing.

**Comment by chris-b1** [comment]
dplyr and SQL would both do this through the 'select` verb, dplyr also has `rename`, which basically works the same way, but selects everything else too.
```
# R
df %>% select(J = a, K = b)
df %>% rename(J = a)  # includes column `b`

# SQL
SELECT a as J, b as K
FROM df
```

In base R it also looks like you can do label assignment via the `names` function (not 100% sure about this)
```
# R
names(df) <- c(J, K)
```

**Comment by chris-b1** [comment]
I suppose #12392 also needs consideration.  From the discussion there, there was some preference towards moving towards a `relabel(labels=['J', 'K'], axis=1)` API, though I personally prefer the named axis arguments for something like this.

cc @nickeubank @jorisvandenbossche @shoyer

**Comment by jorisvandenbossche** [comment]
Yes, we need to take #12392 also in consideration. For now, you only implemented the new feature (renaming all columns using a list-like) in `relabel`, but I think the original idea was to have the full functionality of `rename` in `relabel` (at least, that was my interpretation, but that's certainly open for discussion!). In that case, we should first decide on #12392 before implementing it here.

> though I personally prefer the named axis arguments for something like this.

me as well, but I think there was a compromise in #12392 to have both possibilities at the same time. But let's leave that discussion over there. #12392 is something I would like to see happen for 0.20/1.0, so I can try to take a new look at that this weekend.

**Comment by jreback** [comment]
I view this different than @jorisvandenbossche 

This is *like* ``.set_index`` but w/o the 'guessing' whether you are meaning to set by a particular column (we had a PR try to implement, #11944  this but it got complicated). So I view this as complimentary to ``.rename``, and would have both actually (and keep the API consistent until / unless we change in #12392)

**Comment by jorisvandenbossche** [comment]
Regarding the `relabel` name: I personally find it also not ideal, as `rename(column={'a':'b'})` flows very naturally for "renaming column 'a' to 'b'". 
I think the main problem with `rename` is the conflict with the fact that a Series has a `name` attribute? Or were there other reasons to introduce a `relabel` method?

ibis also calls this relabel: http://docs.ibis-project.org/generated/ibis.expr.api.TableExpr.relabel.html#ibis.expr.api.TableExpr.relabel

cc @pandas-dev/pandas-core

**Comment by jorisvandenbossche** [comment]
> This is like .set_index but w/o the 'guessing' whether you are meaning to set by a particular column

That is certainly also a possibility, but then I would keep `relabel` for exactly that (and not being able to also rename individual labels like `rename`, and then maybe something like `set_labels` would be a clearer name (that you replace the full set of labels)

**Comment by chris-b1** [comment]
>Or were there other reasons to introduce a relabel method?

My thinking was that `rename` is already too overloaded and can run into corner cases with a list-like https://github.com/pandas-dev/pandas/issues/14829#issuecomment-269913291, but can't really be changed for back-compat.

`set_labels` wouldn't be bad either, although somewhat confusingly we have a `set_axis` that does the same thing, but inplace only.

**Comment by TomAugspurger** [comment]
I'm in favor of relabel gaining the dict and function behaviors of rename. I think relabel is the better name for changing the row or column labels. That leaves rename for changing the index and column names.

		_____________________________

**Comment by shoyer** [comment]
The name "labels" is nicely unambiguous about referring to `DataFrame.columns` or `DataFrame.index`, unlike the highly overloaded "name". So all things being equal, `relabel` would seem more consistent with setting index/column values than `rename`.

BUT `rename` already basically does this and I'm not sure a deprecation cycle is worth it. I do think multiple methods may be a good thing, but they should be distinguished by which names they change, not now how they change them (dict vs list).

I am more inclined toward the solution #14829, which enables `list` input to `rename`. I am OK with distinguishing between tuples and lists for dispatching on `Series.name`. If I had my way, `np.array` and `pd.Series` would not coerce tuples (treating them as immutable scalars), only lists.

It's also worth noting that we do have an existing `rename_axis` method, too. To summarize the current state of affairs:
- `Series.rename` sets `Series.index` (`dict`/function) or `Series.name`, depending on the type of input .
- `Series.rename_axis` sets `Series.index` (`dict`/function) or `Series.index.name` (others), depending on the type of input.
- `DataFrame.rename` sets `DataFrame.columns`/`DataFrame.index`
- `DataFrame.rename_axis` sets `DataFrame.columns`/`DataFrame.index` (`dict`/function) or `DataFrame.columns.name`/`DataFrame.index.name` (others), depending on the type of input

CCing @MaximilianR in case he has ideas.

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/issues/14636 here is a proposal to change ``.set_axis`` and make it a full-fledged functions (which is similar to ``.relabel``)

**Comment by chris-b1** [comment]
I'm uncertain about the api here, and this specific change isn't solving a significant problem so closing for now.  Worth re-evaluating after/ as part of #12392.

**Pull request closed without merging**